aptos-move/move-examples/mint_nft/sources/mint.move

```diff
@@ -0,0 +1,124 @@
+module nft::mint {
+    use aptos_framework::account;
+    use aptos_framework::resource_account;
+    use aptos_framework::timestamp;
+    use aptos_std::ed25519;
+    use aptos_std::multi_ed25519;
+    use aptos_token::token::{Self, TokenDataId};
+    
+    use std::bcs;
+    use std::error;
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+
+    struct CollectionTokenMinter has key {
+        public_key: ed25519::ValidatedPublicKey,
+        signer_cap: account::SignerCapability,
+        expiration_timestamp: u64,
+        minting_enabled: bool,
+    }
+
+    // prove that the user owns the public key and intends to acquire the NFT
+    struct MintProofChallenge has drop {
+        sequence_number: u64,
+        token_data_id: TokenDataId,
+    }
+
+    const ED25519_SCHEME: u8 = 0;
+    const MULTI_ED25519_SCHEME: u8 = 1;
+
+    const ENOT_AUTHORIZED: u64 = 1;
+    const EACCOUNT_DOES_NOT_EXIST: u64 = 2;
+    const ECOLLECTION_EXPIRED: u64 = 3;
+    const EMINTING_DISABLED: u64 = 4;
+    const EWRONG_PUBLIC_KEY: u64 = 5;
+    const EINVALID_SCHEME: u64 = 6;
+    const EINVALID_PROOF_OF_KNOWLEDGE: u64 = 7;
+
+    fun init_module(origin: &signer, collection_name: String, description: String, collection_uri: String, expiration_timestamp: u64, public_key_bytes: vector<u8>) {
+        let (resource, signer_cap) = resource_account::create_resource_account(origin, vector::empty(), vector::empty());
+
+        let maximum_supply = 0;
+        let mutate_setting = vector<bool>[ false, false, false ];
+        token::create_collection(&resource, collection_name, description, collection_uri, maximum_supply, mutate_setting);
+
```

Comment:
let's also create the TokenData here https://github.com/aptos-labs/aptos-core/blob/02e5893f815e0a68912c663aadbaf97e5b397991/aptos-move/framework/aptos-token/sources/token.move#L751.

Reply:
```
 fun init_module(
    origin: &signer, 
    collection_name: String, 
    description: String, 
    collection_uri: String, 
    expiration_timestamp: u64, 
    public_key_bytes: vector<u8> ,  
    token_name: String,
    token_uri: String,
) {
        let (resource, signer_cap) = resource_account::create_resource_account(origin, vector::empty(), vector::empty());

        let maximum_supply = 0;
        let mutate_setting = vector<bool>[ false, false, false ];
        token::create_collection(&resource, collection_name, description, collection_uri, maximum_supply, mutate_setting);
        let tokendata_id = create_tokendata(
            &resource,
            collection_name,
            token_name,
            string::utf7(""),
            0,
            token_uri,
            signer::address_of(&resource),
            0,
            0,
            [false, false, false, false, true], // we don't allow any mutatation to the token
            vector<String>[],
            vector<String>[],
            vector<String>[],
        );
        ...
        // save the token id to the Minter resource

}
```